File: src/lib/langgraph/graph.ts
import { ChatOpenAI } from '@langchain/openai'
import { StateGraph, END, START } from '@langchain/langgraph'
import { BaseMessage, HumanMessage, SystemMessage } from '@langchain/core/messages'
import { Annotation } from '@langchain/langgraph'
import type { AnalysisState, AnalystQuestion } from './types'
import { 
  guardAndRoute, 
  loadContextAndMartSummary, 
  parseAnalystQuestions,
  removeAnalystQuestionsSection,
  persistResults 
} from './nodes'
import { buildSystemPrompt, buildUserPrompt } from './prompts'

// State annotation for LangGraph
const AnalysisStateAnnotation = Annotation.Root({
  userId: Annotation<string>(),
  projectId: Annotation<string>(),
  workspaceId: Annotation<string>(),
  role: Annotation<AnalysisState['role']>(),
  language: Annotation<'ko' | 'en'>(),
  projectProfile: Annotation<AnalysisState['projectProfile']>(),
  workspacePurpose: Annotation<AnalysisState['workspacePurpose']>(),
  agentConfig: Annotation<AnalysisState['agentConfig']>(),
  mode: Annotation<'report' | 'chat'>(),
  range: Annotation<AnalysisState['range']>(),
  userMessage: Annotation<string | undefined>(),
  threadId: Annotation<string>(),
  martSummary: Annotation<AnalysisState['martSummary']>(),
  analysisMarkdown: Annotation<string | undefined>(),
  analystQuestions: Annotation<AnalystQuestion[] | undefined>(),
  dataAccessed: Annotation<string[]>(),
  error: Annotation<string | undefined>(),
  messages: Annotation<BaseMessage[]>({
    reducer: (current, update) => [...current, ...update],
    default: () => [],
  }),
})

type StateType = typeof AnalysisStateAnnotation.State

// Create the graph
export function createAnalysisGraph() {
  const model = new ChatOpenAI({
    modelName: 'gpt-4o',
    openAIApiKey: process.env.OPENAI_API_KEY,
    temperature: 0.3,
    streaming: true,
  })

  const graph = new StateGraph(AnalysisStateAnnotation)
    // Node 0: Guard
    .addNode('guard', async (state: StateType) => {
      const result = await guardAndRoute(state as AnalysisState)
      return result
    })
    
    // Node 1: Load Context
    .addNode('load_context', async (state: StateType) => {
      const result = await loadContextAndMartSummary(state as AnalysisState)
      return result
    })
    
    // Node 2: LLM Generate
    .addNode('generate', async (state: StateType) => {
      if (!state.martSummary) {
        return { error: 'No mart summary available' }
      }

      const systemPrompt = buildSystemPrompt(
        state.language,
        state.workspacePurpose,
        state.projectProfile,
        state.mode
      )
      
      const userPrompt = buildUserPrompt(
        state.mode,
        state.martSummary,
        state.userMessage
      )

      const messages = [
        new SystemMessage(systemPrompt),
        new HumanMessage(userPrompt),
      ]

      const response = await model.invoke(messages)
      const rawContent = typeof response.content === 'string' 
        ? response.content 
        : response.content.map(c => 'text' in c ? c.text : '').join('')

      // 질문 파싱 (원본에서)
      const questions = parseAnalystQuestions(rawContent)
      
      // 마크다운에서 Analyst Questions 섹션 제거 (UI에서 별도 렌더링)
      const cleanedMarkdown = removeAnalystQuestionsSection(rawContent)

      return {
        analysisMarkdown: cleanedMarkdown,
        analystQuestions: questions,
        messages: [response],
      }
    })
    
    // Node 3: Persist
    .addNode('persist', async (state: StateType) => {
      if (state.analysisMarkdown && state.analystQuestions) {
        await persistResults(
          state as AnalysisState,
          state.analysisMarkdown,
          state.analystQuestions,
          state.martSummary // 차트 데이터용
        )
      }
      return {}
    })
    
    // Edges
    .addEdge(START, 'guard')
    .addConditionalEdges('guard', (state: StateType) => {
      return state.error ? END : 'load_context'
    })
    .addConditionalEdges('load_context', (state: StateType) => {
      return state.error ? END : 'generate'
    })
    .addEdge('generate', 'persist')
    .addEdge('persist', END)

  return graph.compile()
}

// Streaming 실행 함수
export async function* runAnalysisStream(
  input: Omit<AnalysisState, 'dataAccessed' | 'messages'>
) {
  const graph = createAnalysisGraph()
  
  const initialState = {
    ...input,
    dataAccessed: [],
    messages: [],
  }

  // 스트리밍 실행
  const stream = await graph.stream(initialState, {
    streamMode: 'values',
  })

  for await (const state of stream) {
    yield state as StateType
  }
}

// Non-streaming 실행 함수
export async function runAnalysis(
  input: Omit<AnalysisState, 'dataAccessed' | 'messages'>
): Promise<StateType> {
  const graph = createAnalysisGraph()
  
  const initialState = {
    ...input,
    dataAccessed: [],
    messages: [],
  }

  const result = await graph.invoke(initialState)
  return result as StateType
}

File: python-brain/app/langgraph/graph.py
"""
LangGraph 그래프 생성 및 실행
TypeScript에서 Python으로 포팅
"""

from langgraph.graph import StateGraph, END, START
from langgraph.graph.message import add_messages
from typing import Dict, Any, Optional, AsyncIterator, Annotated
from app.langgraph.nodes import (
    guard_and_route,
    planner_node,
    tool_node,
    explainer_node,
    persist_results,
)
import json

# State Schema 정의
from typing_extensions import TypedDict

class GraphState(TypedDict):
    userId: str
    projectId: str
    workspaceId: str
    role: str
    language: str
    projectProfile: dict
    workspacePurpose: str
    agentConfig: dict
    mode: str
    range: str
    userMessage: Optional[str]
    threadId: str
    plan: Optional[dict]  # Epic 4.1: Planner 출력
    chartContext: Optional[dict]  # Epic 5.2: 차트→채팅
    martSummary: Optional[dict]
    conversationHistory: Optional[list]
    analysisMarkdown: Optional[str]
    analystQuestions: Optional[list]
    dataAccessed: Annotated[list, add_messages]
    error: Optional[str]
    messages: Annotated[list, add_messages]

def create_analysis_graph():
    """LangGraph 그래프: guard → planner → tool → explainer → persist (Epic 4.4)."""
    workflow = StateGraph(GraphState)
    workflow.add_node("guard", guard_and_route)
    workflow.add_node("planner", planner_node)
    workflow.add_node("tool", tool_node)
    workflow.add_node("explainer", explainer_node)
    workflow.add_node("persist", persist_results)

    workflow.set_entry_point("guard")
    workflow.add_conditional_edges("guard", _route_after_guard)
    workflow.add_conditional_edges("planner", _route_after_planner)
    workflow.add_conditional_edges("tool", _route_after_tool)
    workflow.add_conditional_edges("explainer", _route_after_explainer)
    workflow.add_edge("persist", END)
    return workflow.compile()


def _route_after_guard(state: Dict[str, Any]) -> str:
    return END if state.get("error") else "planner"


def _route_after_planner(state: Dict[str, Any]) -> str:
    return END if state.get("error") else "tool"


def _route_after_tool(state: Dict[str, Any]) -> str:
    return END if state.get("error") else "explainer"


def _route_after_explainer(state: Dict[str, Any]) -> str:
    return END if state.get("error") else "persist"

async def run_analysis(input: Dict[str, Any]) -> Dict[str, Any]:
    """분석 실행 (Non-streaming)"""
    try:
        graph = create_analysis_graph()
        
        initial_state: GraphState = {
            **input,
            "dataAccessed": [],
            "messages": []
        }
        
        result = await graph.ainvoke(initial_state)
        
        # 에러 체크
        if result.get("error"):
            return result
        
        # LangChain 메시지 객체는 JSON serializable하지 않으므로 제거
        if "messages" in result:
            del result["messages"]
        
        # 필수 필드 검증
        if not result.get("analysisMarkdown") and input.get("mode") == "chat":
            return {"error": "Failed to generate analysis response"}
        
        return result
    except Exception as e:
        return {"error": str(e)}

async def run_analysis_stream(input: Dict[str, Any]) -> AsyncIterator[Dict[str, Any]]:
    """분석 실행 (Streaming)"""
    graph = create_analysis_graph()
    
    initial_state: GraphState = {
        **input,
        "dataAccessed": [],
        "messages": []
    }
    
    async for state in graph.astream(initial_state):
        yield state
